### PR TITLE
One-hot and thermometer encoding and LSPGA attack

### PR DESCRIPTION
**Pull request by royaurko**
Added the utility functions for discretizing image inputs using one-hot and thermometer encoding, and add the Logit Space-Projected Gradient Ascent (LS-PGA) from the paper https://openreview.net/pdf?id=S18Su--CW. Also added an example tutorial for thermometer encoding and LSPGA.

### CODE REVIEW

**Comment by goodfeli** [comment]
Thanks!

Should I review now or after you've made some updates to get it out of WIP status?

**Comment by royaurko** [comment]
You could review for content/technical correctness, but at the moment this doesn't fit into the Cleverhans spec, hence I tagged it WIP.

**Comment by carlini** [comment]
Can you add some documentation on how to call into these functions? For example, in ls-pga, the model argument appears different than the cleverhans standard model (this one takes a projection function as an argument, and also a reuse?). Similarly, the discretize_fn argument seems like you should be passing discretize_uniform as an argument, but ls-pga doesn't supply a levels argument.

A small example of training a standard model (just plain, without adversarial training) with thermometer encoding, and then showing LS-PGA on it would be a simple way to make this much easier to understand.

**Comment by royaurko** [comment]
I added an ipython example [notebook](https://github.com/royaurko/cleverhans/blob/019fe954f6196bb99f31bd545be16b36c5213075/cleverhans/discretization/Discretization%20Example.ipynb) of how to use the discretization functions and the LS-PGA attack on discretized inputs. Hope that helps!

**Comment by carlini** [comment]
Thanks; but it appears the code has a bug.

On line 40 you call discretize_fn with 1 argument. This is received from line 182, where discretize_fn is an argument to lspga. In the example code you call it with discretize_uniform, which expects at least two arguments, one of which must be levels (and also has a thermometer argument). 

If I change L40 to 
`q = discretize_fn(alpha * low + (1. - alpha) * high)`
then when I train, I end up with the shape of `features_disc_adv` being 192 instead of 24. This appears to be because the shape starts out as (batch_size, 32, 32, 24, 8), presumably incorrectly.

Also, I assume on the ipython notebook that the line
`logits_adv = model(features_adv, is_training=phase)`
should be
`logits_adv = model(features_disc_adv, is_training=phase)`

**Comment by carlini** [comment]
I have been able to make the code into something that runs without crashing. It required changing L40 to:
`q = discretize_fn(alpha * low + (1. - alpha) * high, levels, thermometer)`
as well as ch

**Comment by royaurko** [comment]
You need to set noisy_grads=False while testing, since otherwise you're attacking image + noise and not the image itself. HTH.

**Comment by goodfeli** [comment]
I asked @carlini on hangouts and he says noisy_grads=False successfully reduced the accuracy to 20%

File: cleverhans/discretization/attacks.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+import tensorflow as tf
+import discretization_utils
+
+
+def discretize_range(discretize_fn, levels, low, high, thermometer=False):
+  """Get range of discretized values for in the interval (low, high).
```
Comment (goodfeli) [inline cleverhans/discretization/attacks.py]: I think the word "for" is a typo

File: cleverhans/discretization/Discretization Example.ipynb [hunk 2]
```diff
@@ -0,0 +1,170 @@
+{
```
Comment (goodfeli) [inline cleverhans/discretization/Discretization Example.ipynb]: I think this should go in cleverhans/examples rather than in a new root-level directory

File: cleverhans/discretization/attacks.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+import tensorflow as tf
+import discretization_utils
+
+
+def discretize_range(discretize_fn, levels, low, high, thermometer=False):
```
Comment (goodfeli) [inline cleverhans/discretization/attacks.py]: We should have at least one test for this

File: cleverhans/discretization/attacks.py [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+    # Convert into one hot encoding if q is in thermometer encoding
+    if thermometer:
+      q = discretization_utils.thermometer_to_one_hot(q, levels, flattened=True)
+    out += q
+  return tf.to_float(tf.greater(out, 0.))
+
+
+def adv_dga(x, model, discretize_fn, projection_fn, levels, phase,
+			steps, eps, thermometer=False, noisy_grads=True, y=None):
```
Comment (goodfeli) [inline cleverhans/discretization/attacks.py]: This function doesn't ever seem to be called anywhere. Is it just included for completeness because we describe it in the paper as one of the weaker attacks we tried? If so (If we're just including it as a reference attack implementation) it should be an Attack subclass. If not, maybe we should delete it.

Comment (goodfeli) [reply cleverhans/discretization/attacks.py]: If we keep this function, it should have at least one test

File: cleverhans/discretization/attacks.py [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+	flattened_inp.mask = mask
+    flattened_inp = tf.stop_gradient(flattened_inp)
+
+    cur_x_discretized = flattened_inp
+  return flattened_inp
+
+
+def adv_lspga(x, model, discretize_fn, projection_fn, levels, phase,
+              steps, eps, attack_step=1., thermometer=False,
+              noisy_grads=True, y=None, inv_temp=1., anneal_rate=1.2):
```
Comment (goodfeli) [inline cleverhans/discretization/attacks.py]: should be an Attack subclass

Comment (goodfeli) [reply cleverhans/discretization/attacks.py]: should have at least one test

File: cleverhans/discretization/discretization_utils.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+
+
+def flatten_last(x):
```
Comment (goodfeli) [inline cleverhans/discretization/discretization_utils.py]: Should have at least one test

File: cleverhans/discretization/discretization_utils.py [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+  """
+  shape = x.get_shape().as_list()
+  new_shape = shape[:-1]
+  new_shape[-1] *= shape[-1]
+  new_shape[0] = tf.shape(x)[0]
+  flattened_x = tf.reshape(x, new_shape)
+  return flattened_x
+
+
+def unflatten_last(x, levels):
```
Comment (goodfeli) [inline cleverhans/discretization/discretization_utils.py]: Should have at least one test

File: cleverhans/discretization/discretization_utils.py [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+  discretized_x = one_hot
+  if thermometer:
+    discretized_x = one_hot_to_thermometer(one_hot, levels, flattened=False)
+
+  # Reshape x to [-1, height, width, channels * levels]
+  discretized_x = flatten_last(discretized_x)
+  return discretized_x
+
+
+def undiscretize_uniform(x, levels, flattened=False, thermometer=False):
```
Comment (goodfeli) [inline cleverhans/discretization/discretization_utils.py]: This doesn't seem to actually be called anywhere, do we actually need it?

File: cleverhans/discretization/discretization_utils.py [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+  shape = x.get_shape().as_list()
+  shape[-1] /= levels
+  shape[-1] = int(shape[-1])
+  shape.append(levels)
+  shape[0] = tf.shape(x)[0]
+  unflattened_x = tf.reshape(x, shape)
+  return unflattened_x
+
+
+def discretize_uniform(x, levels, thermometer=False):
```
Comment (goodfeli) [inline cleverhans/discretization/discretization_utils.py]: This doesn't seem to actually be called anywhere, do we actually need it?

File: cleverhans/discretization/discretization_utils.py [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+  discretized_x = one_hot
+  if thermometer:
+    discretized_x = one_hot_to_thermometer(one_hot, levels, flattened=False)
+
+  # Reshape x to [-1, height, width, channels * levels]
+  discretized_x = flatten_last(discretized_x)
+  return discretized_x
+
+
+def get_centroids_by_percentile(x, levels):
```
Comment (goodfeli) [inline cleverhans/discretization/discretization_utils.py]: This doesn't seem to actually be called anywhere, do we actually need it?

File: cleverhans/discretization/discretization_utils.py [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+    return tf.contrib.distributions.percentile(x, q=q, axis=None)
+
+  start = 0.
+  end = 100.
+  quantile_range = tf.lin_space(start, end, levels)
+  centroids = tf.map_fn(quantile, quantile_range)
+  return centroids
+
+
+def discretize_centroids(x, levels, centroids, thermometer=False):
```
Comment (goodfeli) [inline cleverhans/discretization/discretization_utils.py]: This doesn't seem to actually be called anywhere, do we actually need it?

File: cleverhans/discretization/discretization_utils.py [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+    x = unflatten_last(x, levels)
+  if thermometer:
+    int_x = tf.reduce_sum(x, -1) - 1
+  else:
+    int_x = tf.argmax(x, -1)
+  out = tf.to_float(int_x) / (levels - 1)
+  return out
+
+
+def undiscretize_centroids(x,
```
Comment (goodfeli) [inline cleverhans/discretization/discretization_utils.py]: This doesn't seem to actually be called anywhere, do we actually need it?

File: cleverhans/discretization/discretization_utils.py [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+  # [-1, height, width, channels, levels]
+  if flattened:
+    x = unflatten_last(x, levels)
+  if thermometer:
+    x = thermometer_to_one_hot(x, levels, flattened=False)
+  out = tf.reduce_sum(tf.multiply(x, centroids), axis=-1)
+  return out
+
+
+def one_hot_to_thermometer(x, levels, flattened=False):
```
Comment (goodfeli) [inline cleverhans/discretization/discretization_utils.py]: This doesn't seem to actually be called anywhere, do we actually need it?

File: cleverhans/discretization/attacks.py (hunk 4 above)
Comment (carlini) [reply cleverhans/discretization/attacks.py]: It's used if you want to actually discretize the inputs before calling a model.

File: cleverhans/discretization/discretization_utils.py (hunk 9 above)
Comment (carlini) [reply cleverhans/discretization/discretization_utils.py]: It's used if you want to actually discretize the inputs before calling a model.

File: cleverhans/discretization/discretization_utils.py [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+  if flattened:
+    x = unflatten_last(x, levels)
+  thermometer = tf.cumsum(x, axis=-1, reverse=True)
+  # Flatten back if original input was flattened
+  if flattened:
+    thermometer = flatten_last(thermometer)
+  return thermometer
+
+
+def thermometer_to_one_hot(x, levels, flattened=False):
```
Comment (goodfeli) [inline cleverhans/discretization/discretization_utils.py]: This should have at least one test

File: cleverhans/discretization/discretization_utils.py [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+  int_x = tf.to_int32(tf.reduce_sum(x, axis=-1)) - 1
+  one_hot = tf.one_hot(
+      int_x, depth=levels, on_value=1., off_value=0., dtype=tf.float32, axis=-1)
+  # Flatten back if input was flattened
+  if flattened:
+    one_hot = flatten_last(one_hot)
+  return one_hot
+
+
+def random_convolution(x,
```
Comment (goodfeli) [inline cleverhans/discretization/discretization_utils.py]: This doesn't seem to be actually called anywhere, do we actually need it?

**Comment by goodfeli** [comment]
Usually I say "we don't need a tutorial for every new attack / defense, or we'll end up with too many tutorials" but in this case, the extra preprocessing of the input might mean it makes sense to include a tutorial or example showing how the discretization / undiscretization functions should be used in the context of training / attacking a model.

**Comment by royaurko** [comment]
I adapted the code to fit the cleverhans spec, I will add some tests shortly.

**Comment by catherio** [comment]
@goodfeli are you still up to review this? it looks like it's waiting for another review pass.

**Comment by jmgilmer** [comment]
Following up with @catherio and @goodfeli. Looks like it still needs to be reviewed?

**Comment by npapernot** [comment]
Any updates on this PR?

**Comment by goodfeli** [comment]
Sorry, I've consistently been swamped, and keep thinking that's almost over but then it isn't. Is there any chance Jacob or Colin (other thermometer coding co-authors) could review it?

File: cleverhans/attacks.py [hunk 16]
```diff
@@ -1433,3 +1433,106 @@ def generate(self, x, g, **kwargs):
...
+                       thermometer=self.thermometer,
+                       noisy_grads=self.noisy_grads,
+                       inv_temp=self.inv_temp,
+                       y=self.y,
+                       y_target=self.y_target,
+                       anneal_rate=self.anneal_rate)
+
+        return attack
+
+    def parse_params(self, y=None, y_target=None, eps=0.3, steps=40,
```
Comment (craffel) [inline cleverhans/attacks.py]: Not sure if this is common `cleverhans` convention, but a much simpler (and arguably hacky, i.e. it should be a private method) way to write this entire function would be
```python
def parse_params(self, **kwargs):
    self.__dict__.update(kwargs)
```

File: cleverhans/attacks_tf.py [hunk 17]
```diff
@@ -1320,3 +1320,153 @@ def deepfool_attack(sess, x, predictions, logits, grads, sample, nb_candidate,
...
+
+
+def compute_mask(levels, low, high, clip_min=0., clip_max=1.,
+                 thermometer=False):
+    """Get a mask of allowable perturbations in the interval (low, high).
+
+    For example, assume that we uniformly discretize the values
+    between 0 and 1 into 10 bins each represented by either a one hot
+    encoding or a thermometer encoding. Then compute_mask(10, .3, .7)
+    would return [0., 0., 0., 1., 1., 1., 1., 0., 0., 0.]. Note that it's
```
Comment (craffel) [inline cleverhans/attacks_tf.py]: its

File: cleverhans/attacks_tf.py [hunk 18]
```diff
@@ -1320,3 +1320,153 @@ def deepfool_attack(sess, x, predictions, logits, grads, sample, nb_candidate,
...
+    output is independent of the encoding used.
+
+    Args:
```
Comment (craffel) [inline cleverhans/attacks_tf.py]: The docstring format doesn't look consistent across the different functions, is that intentional?

File: cleverhans/attacks_tf.py [hunk 19]
```diff
@@ -1320,3 +1320,153 @@ def deepfool_attack(sess, x, predictions, logits, grads, sample, nb_candidate,
...
+    else:
+        activation_logits = tf.zeros_like(mask)
+
+    for i in range(steps):
+        # Compute one hot representation if input is in thermometer encoding.
+        activation_probs = tf.nn.softmax(
+            inv_temp * (activation_logits * mask - 999999. * (1. - mask)))
+
+        if thermometer:
+            activation_probs = tf.cumsum(activation_probs, axis=-1,
```
Comment (craffel) [inline cleverhans/attacks_tf.py]: Can you use `one_hot_to_thermometer` here?

File: cleverhans_tutorials/mnist_tutorial_lspga.py [hunk 20]
```diff
@@ -0,0 +1,225 @@
+"""
+This tutorial shows how to generate some simple adversarial examples
```
Comment (craffel) [inline cleverhans_tutorials/mnist_tutorial_lspga.py]: Change docstring to mention LSPGA?

File: cleverhans_tutorials/mnist_tutorial_lspga.py [hunk 21]
```diff
@@ -0,0 +1,225 @@
...
+                             input_shape=(None, 28, 28, levels))
+    preds_2 = model_2(x_discretized)
+    lspga2 = LSPGA(model_2, sess=sess)
+    adv_x_2 = lspga2.generate(x, **lspga_params)
+    if not backprop_through_attack:
+        # For the lspga attack used in this tutorial, the attack has zero
+        # gradient so enabling this flag does not change the gradient.
+        # For some other attacks, enabling this flag increases the cost of
+        # training, but gives the defender the ability to anticipate how
+        # the atacker will change their strategy in response to updates to
```
Comment (craffel) [inline cleverhans_tutorials/mnist_tutorial_lspga.py]: attacker

**Comment by npapernot** [comment]
@royaurko Can you make the changes? I'll merge afterwards

**Comment by goodfeli** [comment]
I think let's just close this PR for now. We can re-open it later if we decide to use thermometer / one-hot coding again.
